compiler/src/abi/builder.rs

```diff
@@ -0,0 +1,248 @@
+use crate::abi::elements::{
+    Contract, Event, EventField, FuncInput, FuncOutput, Function, FunctionType, ModuleABIs,
+    VariableType,
+};
+use crate::errors::CompileError;
+use std::collections::HashMap;
+use vyper_parser::ast as vyp;
+
+type TypeDefs<'a> = HashMap<&'a str, &'a vyp::TypeDesc<'a>>;
+type TypeDef<'a> = (&'a str, &'a vyp::TypeDesc<'a>);
+
+/// Parse a map of contract ABIs from the input `module`.
+pub fn module<'a>(module: &'a vyp::Module<'a>) -> Result<ModuleABIs, CompileError> {
+    let type_defs = module
+        .body
+        .iter()
+        .filter(|s| is_type_def(&s.node))
+        .map(|s| type_def(&s.node))
```

Comment:
In places where an `Err` is returned for receiving an incorrect enum variant, it would probably be better to panic (ideally this would be [dealt](https://github.com/rust-lang/rfcs/pull/2593) with at compile time). Result errors should be reserved for things like duplicate type definitions. If an error is returned for something like a duplicate type definition, `abi::build` should return an `Err` to notify the user of a duplicate type definition, so I'm a bit hesitant to filter that out. The other option is to destruct the enum and pass all of its values into the `type_def` function, same goes for all of the other `_def` functions, this way there is no need for a panic. This is the [pattern](https://github.com/g-r-a-n-t/rust-vyper/blob/types/compiler/src/yul/mappers/contracts.rs#L49) I ended up going with in the compiler. It's a bit clunky up-front, but keeps things organized overall. It would also be easy to refactor this into something ideal if enum variant types are introduced to the language.

You mentioned [Scan](https://doc.rust-lang.org/std/iter/struct.Scan.html) before, I'll look into that and see if I can tighten this up a bit. It might be best to just go with the same pattern found in the compiler where state is provided by scan instead of scope.

Reply:
Having thought about this a bit more, it's probably pretty important that we pass the whole `Spanned` object into each mapper function.

e.g. `fn contract_def(scope: Shared<ModuleScope>,  def: Spanned<vyp::ModuleStmt>) -> Result<_, CompilerError>`

This way we can return line numbers and other useful contextual details back to the user if an error occurs. I think this gets back to an earlier point you made about *not throwing away useful information*.

Maybe the statement iterators could look something like this then:

```rust
module
  .body
  .iter()
  .try_fold(HashMap::new(), |mut contracts, stmt| {
    match &stmt.node {
      vyp::ModuleStmt::TypeDef { .. } => type_def(scope, stmt)?,
      vyp::ModuleStmt::SimpleImport { .. } => simple_import(scope, stmt)?,
      vyp::ModuleStmt::FromImport { .. } => from_import(scope, stmt)?,
      vyp::ModuleStmt::ContractDef { .. } => contracts.insert(contract_def(scope, stmt)?),
    };

    Ok(contracts)
  })
  .collect::<Result<HashMap<String, yul::Object>, CompileError>()?
```

It seems pretty clean to me and eliminates the double iteration.

Reply:
if this compiles it looks a lot simpler to me :)

Reply:
Cool, updated this with [code](https://github.com/ethereum/rust-vyper/pull/18/commits/29289b723a104c7bdae6ec739591dbed2d7acb39#diff-0d15d418dfe7b76c76b4203800ee1d02R20) that actually compiles :)

It would probably be a good idea to have some sort of `CompileError` method that formats an error based on a given `Span`, which is something to consider along with this [comment](https://github.com/ethereum/rust-vyper/pull/15#discussion_r421826210). I think it makes sense to create another issue for this.